controller/snapshot_controller.go

```diff
@@ -0,0 +1,584 @@
+	size, err := strconv.ParseInt(snapInfo.Size, 10, 64)
+	if err != nil {
+		return err
+	}
+	snap.Status.Parent = snapInfo.Parent
+	snap.Status.Children = snapInfo.Children
+	snap.Status.MarkRemoved = snapInfo.Removed
+	snap.Status.UserCreated = snapInfo.UserCreated
+	snap.Status.CreationTime = snapInfo.Created
+	snap.Status.Size = size
+	snap.Status.Labels = snapInfo.Labels
+	snap.Status.RestoreSize = restoreSize
+	if snap.Status.MarkRemoved || snap.DeletionTimestamp != nil {
+		snap.Status.ReadyToUse = false
+	} else {
+		snap.Status.ReadyToUse = true
+		snap.Status.Error = ""
+	}
+	return nil
+}
+
+func (sc *SnapshotController) getTheOnlyEngineCRforSnapshot(snapshot *longhorn.Snapshot) (*longhorn.Engine, error) {
+	engines, err := sc.ds.ListVolumeEngines(snapshot.Spec.Volume)
+	if err != nil {
+		return nil, errors.Wrap(err, "getTheOnlyEngineCRforSnapshot")
+	}
+	if len(engines) != 1 {
+		return nil, fmt.Errorf("getTheOnlyEngineCRforSnapshot: found more than 1 engines for volume %v", snapshot.Spec.Volume)
+	}
+	var engine *longhorn.Engine
+	for _, e := range engines {
+		engine = e
+		break
+	}
+	return engine, nil
+}
+
+func (sc *SnapshotController) handleSnapshotCreate(snapshot *longhorn.Snapshot, engine *longhorn.Engine) error {
+	client, err := GetClientForEngine(engine, sc.engineClientCollection, engine.Status.CurrentImage)
+	if err != nil {
+		return err
+	}
+	snapshotInfo, err := client.SnapshotGet(snapshot.Name)
+	if err != nil {
+		return err
+	}
+	if snapshotInfo == nil {
+		sc.logger.Infof("creating snapshot %v of volume %v", snapshot.Name, snapshot.Spec.Volume)
+		_, err = client.SnapshotCreate(snapshot.Name, snapshot.Spec.Labels)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// handleSnapshotDeletion reaches out to engine process to check and delete the snapshot
+func (sc *SnapshotController) handleSnapshotDeletion(snapshot *longhorn.Snapshot, engine *longhorn.Engine) error {
+	client, err := GetClientForEngine(engine, sc.engineClientCollection, engine.Status.CurrentImage)
+	if err != nil {
+		return err
+	}
+	snapshotInfo, err := client.SnapshotGet(snapshot.Name)
+	if err != nil {
+		return err
+	}
+	if snapshotInfo == nil {
+		return nil
+	}
+
+	if !snapshotInfo.Removed {
+		sc.logger.Infof("deleting snapshot %v", snapshot.Name)
+		if err = client.SnapshotDelete(snapshot.Name); err != nil {
+			return err
+		}
+	}
+	// TODO: Check if the purge failure is handled somewhere else
+	purgeStatus, err := client.SnapshotPurgeStatus()
+	if err != nil {
+		return errors.Wrap(err, "failed to get snapshot purge status")
+	}
+	isPurging := false
+	for _, status := range purgeStatus {
+		if status.IsPurging {
+			isPurging = true
+			break
+		}
+	}
+	if !isPurging {
+		sc.logger.Infof("start SnapshotPurge to delete snapshot %v", snapshot.Name)
+		if err := client.SnapshotPurge(); err != nil {
+			return err
+		}
+	}
```

Comment:
I don't think you need to check at caller side, because I would expect the server will check if purging or not. 

https://github.com/longhorn/longhorn-engine/blob/c9ac9c6a3ad2ebd4389a1abf171490c2d8d4aee0/pkg/sync/rpc/server.go#L1185

Reply:
Reaching out to the engine is an expensive operation (the GRPC calls) while checking engine CR is cheap. We don't want to spam the engine process when there are many snapshot CRs being deleted happening near the same time.

Reply:
Good consideration, let's keep it. Thanks.